soco/discovery.py

```diff
@@ -19,7 +22,13 @@
 # pylint: disable=too-many-locals, too-many-branches
 
 
-def discover(timeout=5, include_invisible=False, interface_addr=None):
+def discover(
+    timeout=5,
+    include_invisible=False,
+    interface_addr=None,
+    allow_network_scan=False,
+    min_netmask=24,
```

Comment:
Ok, so I wish I had remembered to mention this when I first suggested this implementation approach. But since we are adding 2 extra arguments to all of them, then we might as well make the second one a `network_scan_kwargs` dict argument and allow full configurability of the network_scan with the same number of arguments. It might also come in handy in the future, since I don't imagine these functions are done growing in size. E.g. we might in the future see them changed to allow not to stop after the first result (to allow to discover S1&S2 systems) and maybe a caching option will be added. What do you think?

Reply:
I'm currently taking  a look at switching args ever to a dict.

Scanning all households (such as the split S1/S2 case) would be a natural enhancement (though we'd probably need to look holistically at SoCo to make sure it operates consistently). Incidentally, that's what I do with the equivalent implementation in soco-cli [1], which also caches the results in a file. I need it myself because I have an S1/S2 system.

[1] https://github.com/avantrec/soco-cli/blob/master/soco_cli/speakers.py

Reply:
This is now implemented in https://github.com/SoCo/SoCo/pull/733/commits/9964b2bc9946e90cacc84483fe3840021b17726c, and tested. I need to check whether the docstrings are correctly formatted, since this is not a pattern I've used much.

Reply:
Hi @pwt.

Ah, I was a little in doubt about how much to write about it. It's, I think, almost the way that it is mostly done.

It is fairly common to collect arguments for a sub-function via the **scan_network, just as you did. I would however not document them there, but simply refer to the function they apply to. Another thing is that `allow_network_scan` really is an argument for the e.g. `discover` function, not for `network_scan`, so I would leave that as a normal argument. Also, this whole thing, with unpacking arguments into a function call from a dict, works just fine even though network scan has normal arguments, so I would definitely keep those, since they carry more information. So to sum up, something like this is what I would suggest, and the way I think that pattern is usually implemented:

```python

def discover(
        timeout=5, include_invisible=False, interface_addr=None,
        allow_network_scan=False, **network_scan_kwargs
):
    """Discover Sonos zones on the local network.

    Return a set of `SoCo` instances for each zone found.
    Include invisible zones (bridges and slave zones in stereo pairs if
    ``include_invisible`` is `True`. Will block for up to ``timeout`` seconds,
     after which return `None` if no zones found.

    Args:
        timeout (int, optional): block for this many seconds, at most.
            Defaults to 5.
        include_invisible (bool, optional): include invisible zones in the
            return set. Defaults to `False`.
        interface_addr (str or None): Discovery operates by sending UDP
            multicast datagrams. ``interface_addr`` is a string (dotted
            quad) representation of the network interface address to use as
            the source of the datagrams (i.e. it is a value for
            `socket.IP_MULTICAST_IF <socket>`). If `None` or not specified,
            the system default interface for UDP multicast messages will be
            used. This is probably what you want to happen. Defaults to
            `None`.
        allow_network_scan (bool, optional): If normal discovery fails, fall
            back to a scan of the attached network(s) to detect Sonos
            devices.
        **network_scan_kwargs: Arguments for the
            :function:`scan_network` function. See its docstring for
            documentation

    Returns: set: a set of `SoCo` instances, one for each zone
        found, or else `None`.

    ...

    """
    ...
    
        elif allow_network_scan:
            _LOG.info("Falling back to network scan discovery")
            return scan_network(
                include_invisible=include_invisible,
                **network_scan_kwargs,
            )

# any_soco similar to by_name below

def by_name(name, allow_network_scan=False, **network_scan_kwargs):
    """Return a device by name.

    Args:
        name (str): The name of the device to return.
        allow_network_scan (bool, optional): If normal discovery fails, fall
            back to a scan of the attached network(s) to detect Sonos
            devices.
        **network_scan_kwargs: Arguments for the
            :function:`scan_network` function. See its docstring for
            documentation Returns: set: a set of `SoCo` instances, one
            for each zone found, or else `None`.

    Returns:
        :class:`~.SoCo`: The first device encountered among all zone with the
            given player name. If none are found `None` is returned.
    """
    devices = discover(
        allow_network_scan=allow_network_scan, **network_scan_kwargs
    )
    if devices is None:
        return None

    for device in devices:
        if device.player_name == name:
            return device
    return None


# Function def args here are as before |
#                         v

# pylint: disable=too-many-statements
def scan_network(max_threads=256, timeout=3.0, min_netmask=24, include_invisible=False):
    """Scan all attached networks for Sonos devices.

    Scans the IPv4 network attached to each interface to check for network devices
    with port 1400 open. Check IPs in parallel threads for efficiency. Once the first
    Sonos device is found, stop checking and use that device to find the other
    devices. Returns a set of `SoCo` instances, or `None` if no Sonos devices are
    discovered.

    This function is intended for use when the usual discovery functions aren't
    working, perhaps due to multicast problems on the network to which the SoCo
    host is attached.

    Args:
        max_threads (int, optional): The maximum number of threads to use when
            scanning the network.
        scan_timeout (float, optional): The network timeout in seconds to use when
            checking each IP address for a Sonos device
        min_netmask (int, optional): The minimum number of netmask bits. Used to
                constrain the network search space.
        include_invisible (bool, optional): Whether to include invisible Sonos devices
            in the set of devices returned.

    Returns:
        set: A set of `SoCo` instances, one for each zone found, or else `None`.
    """
    ...

```

I hope the arguments makes sense, otherwise let me know.

Reply:
Got it. Thanks for the explanation. Amended in https://github.com/SoCo/SoCo/pull/733/commits/449d1668e2abd7ab985707c7af3c09aa76a36192.